go/signedexchange/cbor/encoder.go

```diff
@@ -0,0 +1,233 @@
+	//   0x0a, and never as the bytes 0x5c6e (the characters "\" and "n")
+	//   or as 0x5c7530303061 (the characters "\", "u", "0", "0", "0", and
+	//   "a").
+	//
+	// https://tools.ietf.org/html/rfc7049#section-2.1
+	bs := []byte(s)
+	if !utf8.Valid(bs) {
+		return ErrInvalidUTF8
+	}
+
+	return e.encodeBytes(TypeText, bs)
+}
+
+func (e *Encoder) EncodeArrayHeader(n int) error {
+	// Major type 4:  an array of data items.  Arrays are also called lists,
+	//   sequences, or tuples.  The array's length follows the rules for
+	//   byte strings (major type 2), except that the length denotes the
+	//   number of data items, not the length in bytes that the array takes
+	//   up.  Items in an array do not need to all be of the same type.
+	//   For example, an array that contains 10 items of any type would
+	//   have an initial byte of 0b100_01010 (major type of 4, additional
+	//   information of 10 for the length) followed by the 10 remaining
+	//   items.
+	//
+	// https://tools.ietf.org/html/rfc7049#section-2.1
+	return e.encodeTypedUInt(TypeArray, uint64(n))
+}
+
+func (e *Encoder) encodeMapHeader(n int) error {
+	// Major type 5:  a map of pairs of data items.  Maps are also called
+	//   tables, dictionaries, hashes, or objects (in JSON).  A map is
+	//   comprised of pairs of data items, each pair consisting of a key
+	//   that is immediately followed by a value.  The map's length follows
+	//   the rules for byte strings (major type 2), except that the length
+	//   denotes the number of pairs, not the length in bytes that the map
+	//   takes up.  For example, a map that contains 9 pairs would have an
+	//   initial byte of 0b101_01001 (major type of 5, additional
+	//   information of 9 for the number of pairs) followed by the 18
+	//   remaining items.  The first item is the first key, the second item
+	//   is the first value, the third item is the second key, and so on.
+	//   A map that has duplicate keys may be well-formed, but it is not
+	//   valid, and thus it causes indeterminate decoding; see also
+	//   Section 3.7.
+	return e.encodeTypedUInt(TypeMap, uint64(n))
+}
+
+func (e *Encoder) EncodeBool(b bool) error {
+	ai := byte(0)
+	if b {
+		// True (major type 7, additional information 21)
+		ai = 21
+	} else {
+		// False (major type 7, additional information 20)
+		ai = 20
+	}
+
+	bs := []byte{TypeOther | ai}
+	if _, err := e.w.Write(bs); err != nil {
+		return err
+	}
+	return nil
+}
+
+type MapEntryEncoder struct {
+	keyBuf   bytes.Buffer
+	valueBuf bytes.Buffer
+
+	keyE   *Encoder
+	valueE *Encoder
+}
+
+func NewMapEntry() *MapEntryEncoder {
+	e := &MapEntryEncoder{}
+	e.keyE = &Encoder{&e.keyBuf}
+	e.valueE = &Encoder{&e.valueBuf}
+	return e
+}
+
+func (e *MapEntryEncoder) KeyBytes() []byte {
+	return e.keyBuf.Bytes()
+}
+
+func GenerateMapEntry(f func(keyE *Encoder, valueE *Encoder)) *MapEntryEncoder {
+	e := NewMapEntry()
+	f(e.keyE, e.valueE)
+	return e
+}
+
+func (e *Encoder) EncodeMap(mes []*MapEntryEncoder) error {
```

Comment:
This could be more efficient, but for the data sizes that are likely in signed-exchange maps, it probably doesn't matter. Maybe just comment that?

Reply:
I agree: sorting keys with a new array is not very effective but is the most direct way. I'll comment this later.

Reply:
Done.